File: webhooks/send_webhook_payload.py
#!/usr/bin/env python3
"""Send a webhook payload to the specified Orkes webhook endpoint."""

import requests
from settings import settings


def main():
    """
    Send a webhook payload to the Orkes webhook endpoint.
    """
    headers = {
        "Accept": "application/json",
        "source": settings.source_header,
    }

    # Both the key and list contents must conform to wait_for_webhook.matches
    payload = {
        "user_ids": list(settings.user_ids),
        "type": "customer",
        "agent_input": (
            "Introduce yourself, then inform the users that their emails have been sent."
        ),
    }

    response = requests.post(
        settings.webhook_endpoint_url,
        headers=headers,
        json=payload,
        timeout=30,
    )

    print(f"Status: {response.status_code}")
    print(f"Response: {response.text}")

    # Raise an exception for HTTP error responses (e.g.:4xx and 5xx).
    response.raise_for_status()

    print("Webhook was accepted by Orkes Conductor.")


if __name__ == "__main__":
    main()

File: webhooks/java/src/main/java/io/orkes/codelab/webhooks/SendWebhookPayload.java
package io.orkes.codelab.webhooks;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;

/** Send a webhook payload to the specified Orkes webhook endpoint. */
public final class SendWebhookPayload {

    private SendWebhookPayload() {}

    public static int run() throws IOException, InterruptedException {
        Settings settings = Settings.current();

        Map<String, Object> payload = new LinkedHashMap<>();
        // Exercise 2: Send user_ids instead, the same list as the workflow's input,
        // to match the WAIT_FOR_WEBHOOK task's rule in the deploy-workflow step.
        payload.put("user_id", Settings.USER_ID);
        payload.put("type", "customer");
        // Exercise 3: Replace this prompt with a request to summarize stored email
        // activity, inspect the busiest recipient's history, and return a digest.
        payload.put(
                "agent_input",
                "Introduce yourself, then inform the user that their email has been sent.");
        // Every language version shares one webhook; this key selects this
        // language's workflow (see the matches in the deploy-workflow step).
        payload.put("language", Settings.LANGUAGE);

        HttpRequest request =
                HttpRequest.newBuilder(URI.create(settings.webhookEndpointUrl()))
                        .timeout(Duration.ofSeconds(30))
                        .header("Content-Type", "application/json")
                        .header("Accept", "application/json")
                        .header("source", settings.sourceHeader())
                        .POST(
                                HttpRequest.BodyPublishers.ofString(
                                        new ObjectMapper().writeValueAsString(payload)))
                        .build();

        try (HttpClient client = HttpClient.newHttpClient()) {
            HttpResponse<String> response =
                    client.send(request, HttpResponse.BodyHandlers.ofString());

            System.out.println("Status: " + response.statusCode());
            System.out.println("Response: " + response.body());

            // Throw an exception for HTTP error responses (for example, 4xx and 5xx).
            if (response.statusCode() >= 400) {
                throw new IOException(
                        "Webhook request failed with status " + response.statusCode());
            }
        }

        System.out.println("Webhook was accepted by Orkes Conductor.");
        return 0;
    }
}
